src/btree/bt_sync.c

```diff
@@ -179,115 +175,140 @@ __sync_ref_list_pop(WT_SESSION_IMPL *session, WT_REF_LIST *rlp, uint32_t flags)
 static int
 __sync_ref_obsolete_check(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF_LIST *rlp)
 {
-    WT_ADDR *addr;
-    WT_CELL_UNPACK vpack;
+    WT_ADDR_COPY addr;
+    WT_DECL_RET;
     WT_MULTI *multi;
     WT_PAGE_MODIFY *mod;
-    WT_REF *dup;
-    wt_timestamp_t multi_newest_stop_ts;
-    uint64_t multi_newest_stop_txn;
+    wt_timestamp_t newest_stop_ts;
+    uint64_t newest_stop_txn;
     uint32_t i;
     uint8_t previous_state;
-    bool obsolete;
+    char tp_string[WT_TP_STRING_SIZE];
+    const char *tag;
+    bool busy, hazard, obsolete;
 
     /* Ignore root pages as they can never be deleted. */
     if (__wt_ref_is_root(ref)) {
         __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping root page", (void *)ref);
         return (0);
     }
 
-    /* Lock the ref to avoid any change during the obsolete checks.*/
-    for (;; __wt_yield()) {
-        previous_state = ref->state;
-        if (previous_state != WT_REF_LOCKED &&
-          WT_REF_CAS_STATE(session, ref, previous_state, WT_REF_LOCKED))
-            break;
+    /* Ignore internal pages, these are taken care of during reconciliation. */
+    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL)) {
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping internal page with parent: %p",
+          (void *)ref, (void *)ref->home);
+        return (0);
     }
 
-    /* Ignore deleted pages. */
-    if (previous_state == WT_REF_DELETED) {
+    /* Fast-check, ignore deleted pages. */
+    if (ref->state == WT_REF_DELETED) {
         __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping deleted page", (void *)ref);
-        WT_REF_SET_STATE(ref, WT_REF_DELETED);
         return (0);
     }
 
-    /* Ignore internal pages, these are taken care of during reconciliation. */
-    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL)) {
-        WT_REF_SET_STATE(ref, previous_state);
-        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping internal page with parent: %p",
-          (void *)ref, (void *)ref->home);
+    /* Lock the WT_REF. */
+    WT_REF_LOCK(session, ref, &previous_state);
+
+    /*
+     * If the page is on-disk and obsolete, mark the page as deleted and also set the parent page as
+     * dirty. This is to ensure the parent is written during the checkpoint and the child page
+     * discarded.
+     */
+    newest_stop_ts = WT_TS_NONE;
+    newest_stop_txn = WT_TXN_NONE;
+    obsolete = false;
+    if (previous_state == WT_REF_DISK) {
+        /* There should be an address, but simply skip any page where we don't find one. */
+        if (__wt_ref_addr_copy(session, ref, &addr)) {
```

Comment:
This is the case of `on-page` address and we are forming the `addr` from cell unpack. Can we mention the same in a comment?

Reply:
I don't believe there's any guarantee we're unpacking a cell to acquire the address, that is, the page may be on-disk, but the `WT_REF.addr` structure may be filled out.

For example, if a leaf page was just written and evicted, filling in the `WT_REF.addr`, but the parent page has not yet been itself evicted and read back in, subsequent to the leaf page eviction, we're not going to unpack a cell, we're going to use the in-memory structure.

Please let me know if I'm misunderstanding your comment!

Reply:
Understood.